### PR TITLE
Add FEEvaluation::gather_evaluate and integrate_scatter

### PR DESCRIPTION
**Pull request by kronbichler**
Relates to #5667.

This commit adds two new functions `gather_evaluate` and `integrate_scatter` to `FEEvaluation`. This functions combine `read_dof_values` followed by `evaluate` and `integrate` followed by `distribute_local_to_global`, respectively. In `FEEvaluation` where this new function is introduced right now, this new function at best saves two lines of code because two separate calls can be combined into one, without any change in efficiency.

The real purpose of `gather_evaluate` and `integrate_scatter` will be in `FEFaceEvaluation` where, depending on what gets evaluated, only some of the vector entries must be read. In order to avoid that contribution getting drowned by lots of other changes, I want to put the interface name up to review now. The interface in the `FEFaceEvaluation` class will look exactly the same. 

As to the idea: Think for example of Lagrangian shape functions that have nodes on the face we want to evaluate the values of on the face. In that case, we only need to read the values of the shape functions with support on the face, not all vector values. As reading vectors is often the most costly part in DG, this approach is really useful.

### CODE REVIEW

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -5323,6 +5391,9 @@ ::quadrature_point (const unsigned int q) const
 
 
 
+
+
+
```
Comment (Rombur) [inline include/deal.II/matrix_free/fe_evaluation.h]: Extra blank lines

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 2]
```diff
@@ -1991,16 +1991,84 @@ class FEEvaluation : public FEEvaluationAccess<dim,n_components_,Number>
                  const bool evaluate_gradients,
                  const bool evaluate_hessians = false);
 
+  /**
+   * Evaluates the function values, the gradients, and the Laplacians of the
+   * FE function given at the DoF values in the input array @p values_array at
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: Given at the DoF values -> provided as the DoF values ?

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 3]
```diff
@@ -1991,16 +1991,84 @@ class FEEvaluation : public FEEvaluationAccess<dim,n_components_,Number>
...
+   * the quadrature points on the unit cell. If multiple components are
+   * involved in the current FEEvaluation object, the sorting in values_array
+   * is such that all degrees of freedom for the first component come first,
+   * then all degrees of freedom for the second, and so on. The function
+   * arguments specify which parts shall actually be computed. Needs to be
+   * called before the functions @p get_value(), @p get_gradient() or @p
```
Comment (davydden) [inline include/deal.II/matrix_free/fe_evaluation.h]: > Needs to be called before... give useful information

I would reword it, maybe

This function has to be called first so that... give useful info

@jppelteret can the original be improved or am I just confused?

**Comment by kronbichler** [comment]
@davydden Thanks for the hints, I updated the text regarding your first comment and liked your suggestion in the second one, so I have updated the text accordingly. Please have another look.
